mathics/builtin/forms/base.py

```diff
@@ -0,0 +1,99 @@
+from typing import Callable
+
+import mathics.core.definitions as definitions
+
+from mathics.builtin.base import Builtin
+from mathics.core.element import BaseElement
+from mathics.core.symbols import Symbol
+
+form_symbol_to_class = {}
+
+
+class FormBaseClass(Builtin):
+    """
+    Base class for a Mathics Form.
+
+    All Forms should subclass this.
+    """
+
+    # Mapping from Form and element type to a callable boxing method
+    form_box_methods = {}
+
+    # Using "__new__" is not optimal for what we want.
+    # We basically want to hook into class construction in order to
+    # detect certain class attributes so we can add them to a list.
+    # __new__ has this feature. However we do not really need (or want)
+    # to do the memory allocation aspect that "__new__" is intended for.
+    # We considered __prepare__ and metaclass, instead but could not figure
+    # out how to get that to work.
+    def __new__(cls, *args, **kwargs):
+        """ """
+        instance = super().__new__(cls, expression=False)
+        name = cls.__name__
+
+        if hasattr(cls, "in_printforms") and cls.in_printforms:
+            definitions.PrintForms.add(Symbol(name))
+        if hasattr(cls, "in_outputforms") and cls.in_outputforms:
+            if name in definitions.OutputForms:
+                raise RuntimeError(f"{name} already added to $OutputsForms")
+            definitions.OutputForms.add(Symbol(name))
+        form_symbol_to_class[Symbol(name)] = cls
+        return instance
+
+    @classmethod
+    def box(cls, element: BaseElement, evaluation):
+        """
+        This method is is called for each element that can be boxed.
+        ("box" here is a an active verb, not a noun).
+
+        This is a generic routine which calls the specific boxing routine
+        that has been regstered in class variable ``form_box_methods`` previously.
+
+        If nothing has been registered, we just return ``element`` back unmodified
+        as we do in evaluation.
+
+        Specific and custom methods need to be implemented for each Form
+        and element_type that perform some kind of boxing.
+        """
+        method = cls.form_box_methods.get((cls, element.head), None)
```

Comment:
There is something that I am not seeing in this mechanism: Suppose that we set inside a session a makeboxes rule. Let's say
```
MakeBoxes[F[x_],f_]:=SubscriptBox[RowBox[{"<<",ToString[x],">>"}], "F"]
```

* Where is this new rule stored?
* How `MakeBoxes` find this rule?

Reply:
> There is something that I am not seeing in this mechanism: Suppose that we set inside a session a makeboxes rule. Let's say
> 
> ```
> MakeBoxes[F[x_],f_]:=SubscriptBox[RowBox[{"<<",ToString[x],">>"}], "F"]
> ```
> 
> * Where is this new rule stored?
> * How `MakeBoxes` find this rule?

I am glad you asked this question. Let me direct or attention to [this comment](https://github.com/Mathics3/mathics-core/blob/071da9271b3851479165f7de058d1f52d4f63b50/test/formtest.py#L23-L34) in the demo program.

This code is a sketch. It does not change MakeBoxes. Instead the goal was to give an organization where boxing, and later, formatting decisions have no notion of the form that they are embedded in.

It is possible and possily _likely_ that instead of doing this in methods it needs to be done as rules. In that case instead of the `box_divide()`, `box_power()` methods inside `mathics.builtin.forms.default_boxing`, there would be rules instead. Or maybe this could be totally written as `default_boxing_rules.m` and autoloaded.

Here is a challenge for you if you are up for it.

Using `MakeExpression`, `MakeBoxes`, and `Format` write a new form like `MMateraForm` in WL code. It would simulate one of the non-trivial forms like `StandardForm`.

You are free to "cheat" and use rules from `FormatValues` or any information that an WL implementation will tell you.

Reply:
> I am glad you asked this question. Let me direct or attention to [this comment](https://github.com/Mathics3/mathics-core/blob/071da9271b3851479165f7de058d1f52d4f63b50/test/formtest.py#L23-L34) in the demo program.
> 

OK, I saw that. What I was trying to ask about is how you are envisioning that this could work.



> This code is a sketch. It does not change MakeBoxes. Instead the goal was to give an organization where boxing, and later, formatting decisions have no notion of the form that they are embedded in.
> 
> It is possible and possily _likely_ that instead of doing this in methods it needs to be done as rules. In that case, instead of the `box_divide()`, `box_power()` methods inside `mathics.builtin.forms.default_boxing`, there would be rules instead. Or maybe this could be totally written as `default_boxing_rules.m` and autoloaded.
> 
> Here is a challenge for you if you are up for it...
> 
> Using `MakeExpression`, `MakeBoxes` and `Format`, write a new form like `MMateraForm` in WL code. It would simulate one of the non-trivial forms like `StandardForm`.
> 
> You are free to "cheat" and use rules from `FormatValues` or any information that an WL implementation will tell you.

OK, I could provide you with some examples of that.

Reply:
One other I hope small request here. Please start out with _simple_  rules and Forms that lead to simplicity. There is this tendency to lead off with complicated and convoluted edge-case examples involving features we don't currently support in Mathics. And then we go down a rabbit hole focusing on this which doesn't change the base situation.

Instead, it would be awesome if we could start out with simple rules or Forms mimics that cover the majority of the common cases.

 This is not to say that we won't get to the complicated and convoluted cases, just that we might do that later after we have some simple mastery of doing things more closely as it is done in Mathematica.  (In general, I'd say this is a good thing to practice doing anyway.)